cat/opencl.cat

```diff
@@ -0,0 +1,133 @@
+let rs = mo & rs_prime & ~((mo & ~rs_prime) ; mo)
+
+(* Release-acquire synchronisation *)
+let ra_sw(r) = ((r & [Rel]); Fsb?; [W \ WI]; rs?; r; rf; [R \ WI]; sbF?; ([Acq] & r)) & incl & ~wi
+
+(* Barrier synchronisation *)
+// in OpenCL a barrier results in two fence operations: entry and exit fences:
+// https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_API.html#_work_group_functions
+let bar_sw(r) = (r & [EF]); syncbar & ~wi & swg; (r & [XF])
+
+(* Allowed to synchronise on the other region *)
+let scf = (SC * SC) | ((dynamic_tag(GLOBAL) & dynamic_tag(LOCAL)); unv; (dynamic_tag(GLOBAL) & dynamic_tag(LOCAL)))
+
+(* Global and local synchronises-with *)
+let gsw = ra_sw(dynamic_tag(GLOBAL)) | bar_sw(dynamic_tag(GLOBAL)) | (scf & ra_sw(dynamic_tag(LOCAL)))
+let lsw = ra_sw(dynamic_tag(LOCAL)) | bar_sw(dynamic_tag(LOCAL)) | (scf & ra_sw(dynamic_tag(GLOBAL)))
+
+(******************)
+(* Happens-before *)
+(******************)
+
+(* Global and local happens-before *)
+let ghb = ((dynamic_tag(GLOBAL); sb; dynamic_tag(GLOBAL)) | (dynamic_tag(GLOBAL); IW * ~IW; dynamic_tag(GLOBAL)) | gsw)+
+let lhb = ((dynamic_tag(LOCAL); sb; dynamic_tag(LOCAL)) | (dynamic_tag(LOCAL); IW * ~IW; dynamic_tag(LOCAL)) | lsw)+
+
+irreflexive ghb as O-HbG
+irreflexive lhb as O-HbL
+
+(*************)
+(* Coherence *)
+(*************)
+
+let coh(hb) = (rf^-1)?; mo; rf?; hb
+irreflexive coh(ghb) as O-CohG
+irreflexive coh(lhb) as O-CohL
+
+(************************)
+(* Consistency of reads *)
+(************************)
+
+(* A load can only read from a store that already happened. *)
+irreflexive rf; (ghb | lhb) as O-Rf
+
+(* Visible side effects *)
+let vis(hb) = (W * R) & hb & loc & ~((hb & loc); [W]; hb)
+
+(* A non-atomic load can only read from a store that is visible. *)
+empty (rf; (dynamic_tag(NAL) & dynamic_tag(GLOBAL))) \ vis(ghb) as O-NaRfG
+empty (rf; (dynamic_tag(NAL) & dynamic_tag(LOCAL))) \ vis(lhb) as O-NaRfL
+
+(* Consistency of RMWs *)
+// The original model was tested with Herd, which treats RMW as a single atomic operation.
+// We have modified the model to handle RMW as a sequence of atomic operations.
+// irreflexive rf | (mo;mo;rf^-1) | (mo;rf) as O-Rmw
+empty rmw & (fre; coe) as atomic
+
+(****************************************)
+(* Sequential consistency, simplified,  *)
+(* with scoped SC axioms                *)
+(****************************************)
+
+let scp = Fsb?; (rb | mo | (ghb | lhb)); sbF?
+acyclic (SC*SC) & scp & incl as O-Sscoped
+
+(***************)
+(*    Races    *)
+(***************)
+
+(* data_races *)
+let cnf = ((W * W) | (W * R) | (R * W)) & loc
+let dr = cnf & ~(ghb | lhb) & ~(ghb | lhb)^-1 & ~wi & ~incl
+flag ~empty dr as data_race
+
+(* unsequenced_races *)
+let symm(r) = r | r^-1
+let ur = (wi & cnf & ~symm(po)) \ id
+flag ~empty ur as unsequenced_race
+
+(* Barrier divergence *)
+let bsw = bar_sw(dynamic_tag(GLOBAL)) | bar_sw(dynamic_tag(LOCAL))
+let bd = [EF] & ((~wi & swg); unv) \ (bsw ; unv)
```

Comment:
This looks wild to me. Shouldn't `r;unv` just be `[domain(r)]` here? I mean, the current definition computes huge relations that are then intersected with an identity `[EF]` so all non-identity edges are unnecessary to compute. Thus when right-composing with `unv`, the only relevant edges you get are self-loops which you can directly express using `domain(...)`.

```
let bd = [EF] & [domain(~wi & swg)] \ [domain(bsw)]
```

Reply:
I see, updated